![](img/homework.png)

##### Задание 2: '\w{3}\W\w{10}\W\w{3}'
https://www.hackerrank.com/challenges/matching-word-non-word/problem?isFullScreen=true

##### Задание 1: 
Подсчитайте количество разных слов до и после лемматизации в рамках датасета с классной работы

In [6]:
import itertools
import string
import pandas as pd
import numpy as np
import umap

from matplotlib import pyplot as plt
from IPython.display import clear_output

import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
from nltk.stem import WordNetLemmatizer
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize
from nltk.tokenize import WordPunctTokenizer

ps = PorterStemmer()
lemmatizer = WordNetLemmatizer()

[nltk_data] Downloading package wordnet to /home/vladimir/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /home/vladimir/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [2]:
data = list(open("./quora.txt", encoding="utf-8"))
data[50]

"What TV shows or books help you read people's body language?\n"

In [3]:
tokenizer = WordPunctTokenizer()

print(tokenizer.tokenize(data[50]))

['What', 'TV', 'shows', 'or', 'books', 'help', 'you', 'read', 'people', "'", 's', 'body', 'language', '?']


In [4]:
data_tok = []

for x in data:
    data_tok.append(tokenizer.tokenize(x.lower()))

print(data_tok[1])

['what', 'are', 'some', 'ways', 'to', 'overcome', 'a', 'fast', 'food', 'addiction', '?']


In [5]:
#checking

assert all(isinstance(row, (list, tuple)) for row in data_tok), "please convert each line into a list of tokens (strings)"
is_latin = lambda tok: all('a' <= x.lower() <= 'z' for x in tok)
assert all(map(lambda l: not is_latin(l) or l.islower(), map(' '.join, data_tok))), "please make sure to lowercase the data"


In [19]:
dictionary_before_stem_lem = {}
for i in range(100):
    for j in range(len(data_tok[i])):
        if data_tok[i][j] in dictionary_before_stem_lem.keys():
            dictionary_before_stem_lem[data_tok[i][j]] = dictionary_before_stem_lem[data_tok[i][j]] + 1
        else:
            dictionary_before_stem_lem[data_tok[i][j]] = 1

df_before_stem_lem = pd.DataFrame([dictionary_before_stem_lem]).T

print(f'Количество строк до стемминга и лемматизации: {len(df_before_stem_lem)}')

Количество строк до стемминга и лемматизации: 535


In [10]:
dictionary_before_lem = {}
for i in range(100):
    for j in range(len(data_tok[i])):
      stem = ps.stem(data_tok[i][j])
      if stem in dictionary_before_lem.keys():
          dictionary_before_lem[stem] = dictionary_before_lem[stem] + 1
      else: 
        dictionary_before_lem[stem] = 1

df_before_lem = pd.DataFrame([dictionary_before_lem]).T

print(f'Количество строк до лемматизации: {len(df_before_lem)}')

Количество строк до лемматизации: 513


In [24]:
dictionary_after_lem = {}
for i in range(100):
    for j in range(len(data_tok[i])):
      stem = ps.stem(data_tok[i][j])
      stem = lemmatizer.lemmatize(stem)
      if stem in dictionary_after_lem.keys():
         dictionary_after_lem[stem] = dictionary_after_lem[stem] + 1
      else:
         dictionary_after_lem[stem] = 1

df_after_lem = pd.DataFrame([dictionary_after_lem]).T

print(f'Количество строк после лемматизации: {len(df_after_lem)}')

Количество строк после лемматизации: 512


##### Итого: 
1. всего уникальных слов в первой сотне предложений data_tok составляет 535
2. после стемминга - 513
3. после лемматизации - 512